rfcs/20221006-frontend_event_tracking.md

```diff
@@ -0,0 +1,296 @@
+   * @todo: This potentially requires throttling.
+   **/
+  AUDIO_INTERACTION: {
+    /** The unique ID of the media */
+    id: string;
+    event: "seek" | "play" | "pause";
+    /** The slug (not the prettified name) of the provider */
+    provider: string;
+  };
+  /** The user visits a CC license description page on CC.org  **/
+  VISIT_LICENSE_PAGE: {
+    /** The slug of the license the user clicked on */
+    license: string;
+  };
+  /** The user visits a creator's link in the single result UI  **/
+  VISIT_CREATOR_LINK: {
+    /** The unique ID of the media */
+    id: string;
+    /** The permalink to the creator's profile */
+    url: string;
+  };
+  /** The user right clicks a single image result, most likely to download it. **/
+  RIGHT_CLICK_IMAGE: {
+    /** The unique ID of the media */
+    id: string;
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+  /** The user uses the 'back to search' link on a single result **/
+  BACK_TO_SEARCH: {
+    /** The unique ID of the media */
+    id: string;
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+  /** The visibility of the filter sidebar on desktop is toggled **/
+  TOGGLE_FILTERS: {
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+};
+type AnalyticsEventTypes = keyof AnalyticsEvents;
+
+/**
+ * Send an event to our analytics endpoint. Should be a no-op if the
+ * user has opted-out of analytics.
+ */
+async function sendEvent<T extends AnalyticsEventTypes>(
+  key: T,
+  payload?: AnalyticsEvents[T]
+): Promise<any> {
+  /**
+   * The default values to send with every analytics event.
+   */
+  const defaults = {
+    timestamp: Date.now(),
+    path: window.location.pathname,
+    // The dimensions of the user's device
+    width: window.innerWidth,
+    height: window.innerHeight,
+    referrer: document.referrer || null,
+  };
+  /**
+   * Actually send the event here, with some additional values.
+   * I would have prefered to use `navigator.sendBeacon` here
+   * but it is blocked by most ad blockers.
+   *
+   * We might not want to call our api route "analytics" to
+   * further avoid blocking by adblockers, but I'll use that
+   * name for clarity here.
+   */
+  const response = await fetch(`/analytics`, {
+    method: "POST",
+    headers: { "Content-Type": "application/json" },
+    body: JSON.stringify({ ...defaults, ...payload }),
+  });
+}
+```
+
+### Implementation Plan
+
+This needs more elaboration, as it's a very rough outline. For example, specific events will require more detailed planning.
+
+- [ ] Create "Event Tracking" GitHub milestone in the frontend repository
+- [ ] Design: design a privacy and analytics opt-out checkbox on the Privacy page
+- [ ] Create an issue for the `ANALYTICS_ENABLED` feature flag
+- [ ] Create an issue for the creation of the 'sendEvent' function and for the api route to post analytics events to
```

Comment:
Perhaps will be better to rephrase these in terms of development rather than project management?
> - [ ] Create the `ANALYTICS_ENABLED` feature flag
> - [ ] Create the 'sendEvent' function and the api route to post analytics events to
...

Reply:
Maybe, I do think the issue creation is very important as each event needs to be able to be implemented by someone reading the issue exclusively, who may not have read the RFC. The importance of the issues themselves makes me want to keep track of their creation.

Reply:
I've updated the implementation plan. As you suggested I switched to remove the 'create issue' text, I think you're correct there. More detail has been added too.